### PR TITLE
Vmo 3120 flow lifecycle

### PR DESCRIPTION
**Pull request by jcbashdown**
There a number of TODOs added here which seem like open questions for discussion or things for future tickets to me. Would be good to get your POV.

Embedding in Voto5 is covered here: https://bitbucket.org/voto/voto5/pull-requests/1787/vmo-3120-flow-lifecycle

### CODE REVIEW

File: src/views/Home.vue
Comment (bzabos) [inline src/views/Home.vue:44]: It seems like this piece created the conditions for needing `isConfigured`— Is there a reason we're preferring not to do a full configure here? // How come this `created()` is differs slightly from the others?

File: src/views/InteractionDesigner.vue
Comment (bzabos) [inline src/views/InteractionDesigner.vue:203]: I believe we can get the router to `querystring` serialize our query params for us. Something like this should ensure safely urlencoded param value:
```
const {id: flowId} = this
const {path: nextUrl} = this.$route

this.$router.replace(
    this.route('flows.fetchFlow', {flowId}), 
    {query: {nextUrl}})
```

File: src/views/NewFlow.vue
Comment (bzabos) [inline src/views/NewFlow.vue:16]: Does this need to be a router-link? It seems like we prevent the router-link from doing its natural thing and do our own routing below. Should this just be a vanilla button?

File: src/store/flow/flow.ts
Comment (bzabos) [inline src/store/flow/flow.ts:57]: Out of curiosity: Would you prefer to keep this getter rather than removing + doing a `mapState` directly in the component that consumes `flowsList` ? 

The reason I'm asking is that it feels like a tiny bit of added indirection in a space where we haven't reconciled `IContext`, `IContainer`, `activeFlowContainer` (above) — Might be nice to mitigate having yet another way to find a list of flows 🤷‍♀️ 😸

File: src/components/interaction-designer/toolbar/TreeBuilderToolbar.vue
Comment (bzabos) [inline src/components/interaction-designer/toolbar/TreeBuilderToolbar.vue:39]: Does this end up calling save even when toggling from View Mode into Edit Mode ?

File: stories/CaseBlockStyled.stories.ts
Comment (bzabos) [inline stories/CaseBlockStyled.stories.ts:25]: I think we've already got something like this partially covered from inside of `FlowBuilderSidebarEditorContainer.vue`, like:
```
<flow-builder-container v-if="block || flow">
```

Are we duplicating the logic? Could be nice to not have to remember to add this `v-if` on new block types.

File: src/store/trees/trees.js
Comment (bzabos) [inline src/store/trees/trees.js:61]: This is declared above as `configured: false` — Should this say `ui.configured` or the others be `isConfigured`?

File: src/store/trees/trees.js
Comment (bzabos) [inline src/store/trees/trees.js:55]: Can we rename our `created` property to `isCreated` / `is_created` to indicate boolean? I'd initially read that as a date/timestamp, equivalent to something like `createdAt`. 

`saveCurrentlyInProgress` → `isSaveCurrentlyInProgress`
`configured` → `isConfigured`

**Review by bzabos — CHANGES_REQUESTED** [review]
Hi @jcbashdown, this looks great! I've left a pile of general/non-urgent/context comments, along with 5 suggestions that would be nice to get in: 
- [x] use async/await rather than promises
- [x] use null for absent active flow
- [ ] use bool naming convention + bool value for newly added flags (rather than number)
- [ ] update `activeFlowContainer` to match IContainer
- [ ] log errors into the console in `catch()` blocks

But overall: wow! A big piece we've been talking about for a while, I'm excited to see this settle in as part of the builder's experience 👌

File: src/store/flow/flow.ts
Comment (jcbashdown) [reply src/store/flow/flow.ts:39]: This was me trying to preserve the original type I was finding on master to avoid any weirdness but let me just try and switch to a bool and see how it goes. I think QA will pick up any issues I miss.

File: src/store/flow/flow.ts
Comment (jcbashdown) [reply src/store/flow/flow.ts:57]: Hmm. I think I get what you are saying. The Getter isn't necessary yet so let's move to mapState.

File: src/store/flow/flow.ts
Comment (jcbashdown) [reply src/store/flow/flow.ts:43]: In your checkbox you have "update activeFlowContainer to match IContainer" but what I'm taking from this is that this isn't something we can do quite yet..? It also might not be easy in this PR if we want to avoid conflicts with Safydy's flow-runner upgrade PR perhaps?

File: src/store/flow/flow.ts
Comment (jcbashdown) [reply src/store/flow/flow.ts:144]: I'll switch to label. For the replace, this is taken from what we have for block names (we said we would follow the rules for this in the ticket)

> Update the src/components/interaction-designer/block-editors/LabelEditor.vue component to update the flow name behind the scenes when the label is edited to a string conforming to the same rules as block-editors/NameEditor.

File: src/store/flow/flow.ts
Comment (jcbashdown) [reply src/store/flow/flow.ts]: If it's okay with you, I'd like to leave this for now. Not super familiar with flight monitor and I want to unblock other PRs!

Here's a place holder for this:

https://viamoinc.atlassian.net/browse/VMO-3768

File: src/store/flow/flow.ts
Comment (jcbashdown) [reply src/store/flow/flow.ts]: Done

File: src/components/interaction-designer/toolbar/TreeBuilderToolbar.vue
Comment (jcbashdown) [reply src/components/interaction-designer/toolbar/TreeBuilderToolbar.vue:39]: Good point. I think it does. Fixing this.

File: stories/CaseBlockStyled.stories.ts
Comment (jcbashdown) [reply stories/CaseBlockStyled.stories.ts:25]: For some reason this wasn't quite working for me. Let me try again...

File: src/store/trees/trees.js
Comment (jcbashdown) [reply src/store/trees/trees.js:61]: Good spot. I actually just found a bug where configure was incorrectly being called multiple times. This fixes that!

File: src/store/trees/trees.js
Comment (jcbashdown) [reply src/store/trees/trees.js:55]: Done

File: stories/CaseBlockStyled.stories.ts
Comment (jcbashdown) [reply stories/CaseBlockStyled.stories.ts]: Okay, pushed a fix for this:

https://github.com/FLOIP/flow-builder/pull/35/files#diff-130b7357fda7a41c13a1ec6be92137441c79e5a81a4c8ca6445ad1c1cd75ae9fR75

This crashes if no flow is yet present. Slightly unclear why this wasn't an issue before...

**Comment by jcbashdown** [comment]
@bzabos I believe I've not addressed all your comments. I've found one small flow switching error while doing this which I'm looking at now. Other than that I'm hoping this is enough to move to QA. Please let me know if I missed or misunderstood something though!

File: src/store/flow/flow.ts
Comment (jcbashdown) [inline src/store/flow/flow.ts:54]: This will become IContainer when the flow-runner version bump is merged. These TODO fields all need to be deleted for now based on discussion in backend runtime hilda. CC @bzabos . I'll add in the latest spec version now it's settled a bit. Is there anywhere good we can pull this from dynamically?

Comment (jcbashdown) [reply src/store/flow/flow.ts:54]: - Will do this in when moving to IContainer after this is merged to master and the flow-runner version bump is also merged there.

File: src/views/NewFlow.vue
Comment (jcbashdown) [inline src/views/NewFlow.vue:47]: @bzabos Found a bug with this today that I've been wrestling with but I think I've not decided to leave fixing for now. Basically, when you hit this page a blank flow is added to the container. When you "Save and continue" we hit the backend dummy route for persisting and move to the interaction designer. The issue is, if you came to this page from the link on Home.vue and then instead of submitting you go back to the homepage with the back button (which vue-router handles) you will see the empty flow in the list there - even though you never saved it.

I think a fix for this is a bit tricky as all of the flow-editing components edit the currently active flow which comes from the active flow container. It would be a big-ish refactor to let these components edit this or alternatively edit a temporary flow that is not yet in the container. This might also result in a pretty ugly component. 

The alternative is to set some sort of isPersisted/isCreated flag on the flows when we persist and then not render un persisted/created flows on the home page. Then we can find some point in the flow lifecycle to clear out unpersisted flows. This feels a bit hacky though.

As I understand it, we may in future have a draft state that we can use here and do something like option 1 in a more though out way. This coupled with the fact that this Home component won't actually ever be used by voto5 (the flows index will end up coming from php there and any navigation to this page will pull only the actually persisted flows) means I don't think this is a big enough issue to invest time in fixing now. Let me know if you disagree though.

File: src/views/FetchFlow.vue
Comment (safydy) [inline src/views/FetchFlow.vue:31]: @jcbashdown , can you try removing this `<any>` for new added components, I think it won't be necessary. (For existing component, I removed those `<any>` on vmo-2623 PR https://github.com/FLOIP/flow-builder/pull/55/files)

File: src/store/flow/flow.ts
Comment (bzabos) [reply src/store/flow/flow.ts]: Ohh, I see! From my perspective, it would be really nice to see "Hello quick brown fox" result in "hello_quick_brown_fox", I don't know what ticket that would be a part of, though. If it doesn't feel like it should be a part fo this changeset, that's also sounds totally fine to me.

File: src/store/flow/flow.ts
Comment (bzabos) [reply src/store/flow/flow.ts]: Sounds great!

File: src/views/FetchFlow.vue
Comment (bzabos) [reply src/views/FetchFlow.vue:31]: CC/ @poojahpatil90

File: src/views/NewFlow.vue
Comment (bzabos) [reply src/views/NewFlow.vue:46]: Nice! To your conclusion: I agree with you, no need to fix for now! Can we create a JIRA to revisit this in `2.0`-ish?

Overall, I'm partial to the `draft` concept where we would edit a local flow, then hitting _Continue_ would then add the flow to the state. 

Begs the question of whether or not we should update our approach to vuex actions API to be more of a functional toolset like lodash, eg: `setFlowName("My First Flow", {flow})` which would then modify the _reference_ (rather than performing lookup from state) while simultaneously continuing to provide an encapsulation of any logic needed behind the scenes.

File: stories/CaseBlockStyled.stories.ts
Comment (bzabos) [reply stories/CaseBlockStyled.stories.ts]: This link doesn't end up landing me on anything for me 😛 !

Comment (bzabos) [reply stories/CaseBlockStyled.stories.ts]: I think my question is whether we should update the guard that already exists and would only need a change in one place, rather than adding an additional guard to every story?

File: src/store/flow/flow.ts
Comment (bzabos) [reply src/store/flow/flow.ts:43]: My suggestion was to update the set of properties listed below to match the properties on flow runner's `IContainer` to pave way for when we do have access to `IContainer`.

If you're suggesting that this affects more than just a couple places, then yes let's leave it for now!

**Review by bzabos — APPROVED** [review]

**Comment by bzabos** [comment]
@jcbashdown I've switch my review to Approved and left a couple last comments. I'm leaving it up to your discretion as to how to handle those. Thanks for taking the time to respond to the feedback!

File: src/store/flow/flow.ts
Comment (jcbashdown) [reply src/store/flow/flow.ts:43]: Interesting. I was basing this on the spec. I'll take a quick look at the differences from IContainer

Comment (jcbashdown) [reply src/store/flow/flow.ts:43]: Hmm. I can't find IContainer. Where is this..? Not seeing it in the latest flow-runner.

File: stories/CaseBlockStyled.stories.ts
Comment (jcbashdown) [reply stories/CaseBlockStyled.stories.ts]: Yup, exactly. Not sure why that link doesn't load. Let me try again:

https://github.com/FLOIP/flow-builder/pull/35/files#diff-130b7357fda7a41c13a1ec6be92137441c79e5a81a4c8ca6445ad1c1cd75ae9fR85

That link works for me to show what I needed to make the existing guard work. This allowed me to delete all changes to stories I had made previously which you can now see in the diff (which no longer includes any stories)

File: src/views/Home.vue
Comment (jcbashdown) [reply src/views/Home.vue:44]: This is just a shortcut really. To be honest it's probably a bit lazy. We decided that this component is only used in the builder and intended to be overridden when using the app. I actually think that won't always be the case. I'm going to add in a full configure now.

**Pull request merged**
